db/db.js

```diff
@@ -238,27 +238,71 @@ async function updateSentAlerts(locationid, sentalerts, clientParam) {
   }
 }
 
-// Saves the state and incident type into the sessions table
-async function saveAlertSession(chatbotState, incidentType, sessionid, clientParam) {
+async function saveSession(session, clientParam) {
   try {
+    const results = await runQuery(
+      'saveSessionSelect',
+      `
+      SELECT *
+      FROM sessions
+      WHERE id = $1
+      LIMIT 1
+      `,
+      [session.id],
+      clientParam,
+    )
+    if (results === undefined || results.rows === undefined || results.rows.length === 0) {
+      throw new Error("Tried to save a session that doesn't exist yet. Use createSession() instead.")
+    }
+
     await runQuery(
-      'saveAlertSession',
-      'UPDATE sessions SET chatbot_state = $1, incident_type = $2 WHERE id = $3',
-      [chatbotState, incidentType, sessionid],
+      'saveSessionUpdate',
+      'UPDATE sessions SET locationid = $1, state = $2, phone_number = $3, notes = $4, incident_type = $5, chatbot_state = $6, alert_type = $7, responded_at = $8 where id = $9',
+      [
+        session.locationid,
+        session.state,
+        session.phoneNumber,
+        session.notes,
+        session.incidentType,
+        session.chatbotState,
+        session.alertType,
+        session.respondedAt,
+        session.id,
+      ],
       clientParam,
     )
   } catch (err) {
     helpers.log(JSON.stringify(err))
   }
 }
 
-// Saves the state and incident type into the sessions table
-async function updateSession(id, clientParam) {
+async function getHistoricAlertsByAlertApiKey(alertApiKey, maxHistoricAlerts, maxTimeAgoInMillis, clientParam) {
   try {
-    await runQuery('updateSession', 'UPDATE sessions SET updated_at = now() WHERE id = $1', [id], clientParam)
+    // Historic Alerts are those with status "Completed" or that were last updated longer ago than the SESSION_RESET_TIMEOUT
+    const results = await runQuery(
+      'getHistoricAlertsByAlertApiKey',
+      `
+      SELECT s.id, l.display_name, s.incident_type, s.alert_type, s.created_at, s.responded_at
+      FROM sessions AS s
+      LEFT JOIN locations AS l ON s.locationid = l.locationid
+      WHERE l.alert_api_key = $1
+      AND (
+        s.chatbot_state = $2
+        OR s.updated_at < now() - $3 * INTERVAL '1 millisecond'
+      )
+      ORDER BY s.created_at DESC
+      LIMIT $4
+      `,
+      [alertApiKey, ALERT_STATE.COMPLETED, maxTimeAgoInMillis, maxHistoricAlerts],
+      clientParam,
+    )
+
+    return results.rows
   } catch (err) {
     helpers.log(JSON.stringify(err))
   }
+
+  return null
```

Comment:
So I noticed we have an expanding variety of behaviors when pg returns something other than the array of query results we expect. Throughout db.js we're doing:

`    if (results === undefined || results.rows.length === 0) {
      return null
    }
`

`    if (results === undefined) {
      helpers.log('Error: No location with associated API key exists')
      return null
    }
`

`    if (!results) {
      return null
    }
`

and now this - where null is default if the try block errors out (but not if it returns an empty array!). Seems worth it to decide on one? I think this one makes sense, and then the callers just need to be smart about handling empty arrays (which in most case works fine?)

Reply:
Just checked. In this case the caller is fine with an empty array. So we only want it to return null on errors.